traffic_ops/traffic_ops_golang/login/login.go

```diff
@@ -104,3 +108,119 @@ func LoginHandler(db *sqlx.DB, cfg config.Config) http.HandlerFunc {
 		fmt.Fprintf(w, "%s", respBts)
 	}
 }
+
+func OauthLoginHandler(db *sqlx.DB, cfg config.Config) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		handleErrs := tc.GetHandleErrorsFunc(w, r)
+		defer r.Body.Close()
+		authenticated := false
+		resp := struct {
+			tc.Alerts
+		}{}
+
+		form := auth.PasswordForm{}
+		tokenForm := struct {
+			Token string `json:"t"`
+		}{}
+
+		if err := json.NewDecoder(r.Body).Decode(&tokenForm); err != nil {
+			handleErrs(http.StatusBadRequest, err)
+			return
+		}
+
+		encodedToken := tokenForm.Token
+
+		if encodedToken == "" {
+			log.Errorf("Token not found in request but is required")
+			handleErrs(http.StatusBadRequest, errors.New("Token not found in request but is required"))
+			return
+		}
+
+		decodedToken, err := jwt.Parse(encodedToken, func(unverifiedToken *jwt.Token) (interface{}, error) {
+			publicKeyUrl := unverifiedToken.Header["jku"].(string)
+			publicKeyId := unverifiedToken.Header["kid"].(string)
+
+			if !VerifyUrlOnWhiteList(publicKeyUrl, cfg.ConfigTrafficOpsGolang.WhitelistedOAuthUrls) {
+				return nil, errors.New("Key URL from token is not included in the whitelisted urls. Received: " + publicKeyUrl)
+			}
+
+			keys, err := jwk.FetchHTTP(publicKeyUrl)
+			if err != nil {
+				return nil, err
+			}
+
+			keyById := keys.LookupKeyID(publicKeyId)
+			selectedKey, err := keyById[0].Materialize()
+
+			if err != nil {
+				return nil, err
+			}
+
+			return selectedKey, nil
+		})
+		if err != nil {
+			handleErrs(http.StatusInternalServerError, errors.New("Error decoding token with message: "+err.Error()))
```

Comment:
I was torn on this.  Since the KeyFunc is passed into the Parse method, this line will catch all of the errors above, including errors from Materialize, bad key set URLs, bad public key, etc.  So I'm open to suggestions if InternalServerError is not appropriate, but this error is a catch-all for the other errors

Reply:
Yeah, looks like I misunderstood what was happening here. If you want, you could call `handleErrs` _within_ the anonymous `func` and set a boolean variable local to the caller to indicate if a returned `error` was internal to the anonymous `func`. Something like:

```go
internal := false
decodedToken, err := jwt.Parse(encodedToken, func(unverifiedToken *jwt.Token) (interface{}, error) {
    thing, err := funcThatCouldErr()
    if err != nil {
        internal = true
        handleErrs(http.ApropriateStatus, errors.New("Error description"))
        return nil, err
    }
    // ...
}
if internal {
    return
} else if err != nil {
    handleErrs(http.StatusInternalServerError, errors.New("Error parsing JWT: "+err.Error())
    return
}
```

Reply:
But idk how big a deal that is - I don't know of any ATC client sophisticated enough to care about the appropriateness of the status code. Or any human (besides me) that cares. If you want to leave it the way you have it, that's fine with me.

Reply:
i kinda think if anything fails in the parsing of this token, just return a 401?

401 means "sorry, don't know you you are." you currently get that if:

1 - username and/or password is not found so TO doesn't know who you are
2 - mojo cookie is expired so TO doesn't know who you are

this kinda looks like #2